File: Leetcode/993.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    static const int MAXN = 105;
    int node2height[MAXN];
    int node2parent[MAXN];
    void dfs(TreeNode* root, int currHeight, TreeNode* fa) {
        node2height[root->val] = currHeight;
        if (fa != nullptr) {
            node2parent[root->val] = fa->val;
        }
        if (root->left && root->right) {
            dfs(root->left, currHeight + 1, root);
            dfs(root->right, currHeight + 1, root);
        }
        else if (root->left && !root->right) {
            dfs(root->left, currHeight + 1, root);
        }
        else if (!root->left && root->right) {
            dfs(root->right, currHeight + 1, root);
        }
    }
    bool isCousins(TreeNode* root, int x, int y) {
        if (!root) return false;
        node2parent[root->val] = -1;
        dfs(root, 0, nullptr);
        return node2height[x] == node2height[y] && node2parent[x] != node2parent[y];
    }
};

File: Leetcode/993/993.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
const MAXN = 105
var depth [MAXN]int
var parent [MAXN]int

func dfs1(root *TreeNode, d int) {
	if root != nil {
		depth[root.Val] = d
		dfs1(root.Left, d+1)
		dfs1(root.Right, d+1)
	}
}

func dfs2(u, fa *TreeNode) {
	if u != nil {
		if fa != nil {
			parent[u.Val] = fa.Val
		}
        dfs2(u.Left, u)
        dfs2(u.Right, u)
	}
}

func isCousins(root *TreeNode, x int, y int) bool {
	dfs1(root, 0)
	parent[root.Val] = -1
	dfs2(root, nil)
	return depth[x] == depth[y] && parent[x] != parent[y]
}